ipa-core/src/helpers/transport/stream/mod.rs

```diff
@@ -42,6 +46,144 @@ pub type BoxBytesStream = Pin<Box<dyn BytesStream>>;
 //  * Avoiding an extra level of boxing in the production configuration using axum, since
 //    the axum body stream type is already a `Pin<Box<dyn HttpBody>>`.
 #[cfg(feature = "in-memory-infra")]
-pub type BodyStream = WrappedBoxBodyStream;
+type BodyStreamInner = WrappedBoxBodyStream;
 #[cfg(feature = "real-world-infra")]
-pub type BodyStream = WrappedAxumBodyStream;
+type BodyStreamInner = WrappedAxumBodyStream;
+
+/// Wrapper around [`BodyStreamInner`] that enforces checks relevant to both in-memory and
+/// real-world implementations.
+pub struct BodyStream {
+    inner: BodyStreamInner,
+}
+
+impl BodyStream {
+    pub fn new(bytes: Bytes) -> Self {
+        let stream = iter(bytes.split().into_iter().map(Ok::<_, BoxError>));
+        Self::from_bytes_stream(stream)
+    }
+
+    #[must_use]
+    pub fn empty() -> Self {
+        Self {
+            inner: BodyStreamInner::empty(),
+        }
+    }
+
+    pub fn from_bytes_stream(stream: impl BytesStream + 'static) -> Self {
+        Self {
+            inner: BodyStreamInner::from_bytes_stream(stream),
+        }
+    }
+}
+
+impl Stream for BodyStream {
+    type Item = Result<Bytes, BoxError>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let next = self.inner.poll_next_unpin(cx);
+        if let Poll::Ready(Some(Ok(v))) = &next {
+            debug_assert!(
+                v.len() <= MAX_HTTP_CHUNK_SIZE,
+                "Chunk size {} is greater than maximum allowed {MAX_HTTP_CHUNK_SIZE} bytes",
+                v.len()
+            );
+        };
+
+        next
+    }
+}
+
+impl From<Vec<u8>> for BodyStream {
+    fn from(value: Vec<u8>) -> Self {
+        Self::new(Bytes::from(value))
+    }
+}
+
+#[cfg(feature = "web-app")]
+#[async_trait::async_trait]
+impl<S> axum::extract::FromRequest<S> for BodyStream
+where
+    S: Send + Sync,
+{
+    type Rejection = crate::net::Error;
+
+    async fn from_request(req: axum::extract::Request, state: &S) -> Result<Self, Self::Rejection> {
+        Ok(Self {
+            inner: BodyStreamInner::from_request(req, state).await?,
+        })
+    }
+}
+
+/// The size is chosen somewhat arbitrary - feel free to change it, but don't go above 2Gb as
+/// that will cause Hyper's HTTP2 to fail.
+const MAX_HTTP_CHUNK_SIZE: usize = 1024 * 1024; // 1MB
+const_assert!(MAX_HTTP_CHUNK_SIZE > 0);
```

Comment:
would it be worth adding a check that makes sure it doesn't go over 2gb pointing to the hyper issue in github?

Reply:
good point, will add